src/platform/tls_schannel.c

```diff
@@ -816,221 +808,165 @@ const static UNICODE_STRING QuicTlsPackageName = RTL_CONSTANT_STRING(L"Schannel"
-    TLS_PARAMETERS LocalTlsParameters = { 0 };
-    PSCH_CREDENTIALS Credentials = &LocalCredentials;
-    Credentials->pTlsParameters = &LocalTlsParameters;
-    Credentials->cTlsParameters = 1;
-#endif
-
-    //
-    // Initialize user/kernel-common configuration.
-    //
-    Credentials->dwVersion = SCH_CREDENTIALS_VERSION;
-    Credentials->pTlsParameters->grbitDisabledProtocols = (DWORD) ~SP_PROT_TLS1_3_SERVER;
-    Credentials->pTlsParameters->cAlpnIds = 0;
-    Credentials->pTlsParameters->rgstrAlpnIds = NULL; // QUIC manages all the ALPN matching.
-    Credentials->pTlsParameters->cDisabledCrypto = 0;
-    //
-    // TODO: Disallow AES_CCM_8 algorithm, which are undefined in the QUIC-TLS spec.
-    //
-    Credentials->pTlsParameters->pDisabledCrypto = NULL;
-    Credentials->dwFlags |= SCH_CRED_NO_SYSTEM_MAPPER;
-
-    //
-    // This flag disables known-weak crypto algorithms.
-    //
-    Credentials->dwFlags |= SCH_USE_STRONG_CRYPTO;
-
-    if (Flags & QUIC_SEC_CONFIG_FLAG_ENABLE_OCSP) {
-        Credentials->dwFlags |= SCH_CRED_SNI_ENABLE_OCSP;
-    }
-
-    //
-    // Kernel mode-only APIs and configuration.
-    //
-#ifdef _KERNEL_MODE
-
-    Config->SspiContext = SspiCreateAsyncContext();
-    if (Config->SspiContext == NULL) {
-        QuicTraceEvent(
-            LibraryError,
-            "[ lib] ERROR, %s.",
-            "SspiCreateAsyncContext");
-        Status = QUIC_STATUS_OUT_OF_MEMORY;
-        goto Error;
-    }
-
-    SecStatus =
-        SspiSetAsyncNotifyCallback(
-            Config->SspiContext,
-            QuicTlsSspiNotifyCallback,
-            AchaContext);
-    if (SecStatus != SEC_E_OK) {
-        QuicTraceEvent(
-            LibraryErrorStatus,
-            "[ lib] ERROR, %u, %s.",
-            SecStatus,
-            "SspiSetAsyncNotifyCallback");
-        Status = SecStatusToQuicStatus(SecStatus);
+    QUIC_ACH_CONTEXT* AchContext =
+        QuicTlsAllocateAchContext(
+            CredConfig,
+            Context,
+            CompletionHandler,
+            Config);
+    if (AchContext == NULL) {
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
         goto Error;
     }
 
+    PSCH_CREDENTIALS Credentials = &AchContext->Credentials;
+    Credentials->pTlsParameters = &AchContext->TlsParameters;
+    Credentials->cTlsParameters = 1;
+
+    Credentials->dwVersion = SCH_CREDENTIALS_VERSION;
+    Credentials->dwFlags |= SCH_CRED_NO_SYSTEM_MAPPER;
+    Credentials->dwFlags |= SCH_USE_STRONG_CRYPTO;
+    if (CredConfig->Flags & QUIC_CREDENTIAL_FLAG_NO_CERTIFICATE_VALIDATION) {
+        Credentials->dwFlags |= SCH_CRED_MANUAL_CRED_VALIDATION;
+    }
+    if (CredConfig->Flags & QUIC_CREDENTIAL_FLAG_ENABLE_OCSP) {
+        Credentials->dwFlags |= SCH_CRED_SNI_ENABLE_OCSP;
+    }
+    if (CredConfig->Flags & QUIC_CREDENTIAL_FLAG_CLIENT) {
+        Credentials->dwFlags |= SCH_CRED_NO_DEFAULT_CREDS;
+        Credentials->pTlsParameters->grbitDisabledProtocols = (DWORD)~SP_PROT_TLS1_3_CLIENT;
```

Comment:
If you're using the same configuration for server and client, this might cause problems

Reply:
They can't be the same